Add content assertion XML tags for test output verification using images

kostrykin opened the pull request:
As part of the Imaging Hackathon in Freiburg. This is a follow-up PR of #17556.

## Contributions

This PR adds content assertion XML tags for automated test output verification using 2-D single-channel and multi-channel images. The contributed XML tags are detailed below.

### `image_has_metadata`

The `image_has_metadata` XML tag can be used for verification of the **metadata** of the images generated by a Galaxy tool wrapper. Currently supported metadata is:

| Attribute | Feature description |
| --- | --- |
| `width` | The width of the image. |
| `height` | The height of the image. |
| `channels` | Number of channels. |

Example:
```xml
<test>
  <output name="output">
    <assert_contents>
      <image_has_metadata width="512" height="512" channels="1" />
    </assert_contents>
  </output>
</test>
```

### `image_has_intensities`

The `image_has_intensities` XML tag can be used for verification of the **image intensities** of the images generated by a Galaxy tool wrapper. Verification of the following features is supported:

| Attribute | Feature description |
| --- | --- |
| `mean_intensity` | The mean image intensity. |
| `mean_intensity_min` | Lower bound of the mean image intensity. |
| `mean_intensity_max` | Upper bound of the mean image intensity. |
| `center_of_mass` | The center of mass of the image intensities. |
| `eps` | Absolute tolerance for `mean_intensity` and `center_of_mass` assertions. |
| `channel` | Restricts the assertions to only a specific image channel. |

The assertions are imposed for all image channels if the `channel` attribute is omitted.

Example:
```xml
<test>
  <output name="output">
    <assert_contents>
      <image_has_intensities channel="0" mean_intensity="0.24" eps="0.01" />
      <image_has_intensities channel="1" mean_intensity="0.66" eps="0.01" />
    </assert_contents>
  </output>
</test>
```

### `image_has_labels`

The `image_has_labels` XML tag can be used for verification of **label images (label maps)** generated by a Galaxy tool wrapper. Verification of the following features is supported:

| Attribute | Feature description |
| --- | --- |
| `number_of_objects` | The number of uniquely labeled objects. |
| `mean_object_size` | The mean size (number of pixels) of uniquely labeled objects. |
| `mean_object_size_min` | Lower bound of the mean size of uniquely labeled objects. |
| `mean_object_size_max` | Upper bound of the mean size of uniquely labeled objects. |
| `exclude_labels` | Excludes labels from the other assertions. |
| `eps` | Absolute tolerance for the `mean_object_size` assertion. |

The primary use case of the `exclude_labels` attribute is to exclude the image background from the `number_of_objects` and `mean_object_size` assertions, which is often labeled by `0` or `-1`.

Example:
```xml
<test>
  <output name="output">
    <assert_contents>
      <image_has_labels number_of_objects="187" exclude_labels="0" />
    </assert_contents>
  </output>
</test>
```

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

kostrykin committed:
Add `has_image_metadata` test assertion

kostrykin committed:
Fix test

kostrykin committed:
Fix syntax error

kostrykin committed:
Add missing import

kostrykin committed:
Fix test

kostrykin committed:
Fix bug

kostrykin committed:
Fix bug

kostrykin committed:
Add more metadata assertions

kostrykin committed:
Add tests for label images

kostrykin committed:
Fix XML

kostrykin committed:
Fix XML

kostrykin committed:
Fix bug

kostrykin committed:
Fix bug

kostrykin committed:
Fix bug

kostrykin committed:
Fix tests

kostrykin committed:
Add `image_has_intensities` and tests

kostrykin committed:
Fix bug

kostrykin committed:
Fix tests

kostrykin committed:
Fix tests

kostrykin committed:
Add debug info

kostrykin committed:
Fix bug

kostrykin committed:
Fix tests

kostrykin committed:
Add test

kostrykin committed:
Add XSD for `image_has_metadata`

kostrykin committed:
Fix bug in test

kostrykin committed:
Update XSD

kostrykin committed:
Update XSD

kostrykin committed:
Fix assertion errors

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Test

kostrykin committed:
Test

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Add missing type checks

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

kostrykin committed:
Fix linting issues

bernt-matthias reviewed (commented):
Super cool! 

Maybe it would be good to reuse this scheme for asserting numeric values https://github.com/galaxyproject/galaxy/blob/cef07534d54b6e0ad55623534f909f446fb0a1c0/lib/galaxy/tool_util/verify/asserts/size.py#L22C5-L22C19, i.e. use `delta`, `min` and `max`?

kostrykin committed:
Add `*_min` and `*_max` for `mean_intensity` and `mean_object_size`, fix default value for `eps`

- Add `mean_intensity_min` and `mean_intensity_max` attributes for `image_has_intensities` assertion tag
- Add `mean_object_size_min` and `mean_object_size_max` attributes for `image_has_labels` assertion tag
- Fix default value for `eps`.

kostrykin committed:
Add `_assert_float` function and refactor

bernt-matthias commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
````diff
@@ -2739,6 +2751,123 @@ $attribute_list::5
+Asserts the labels in a label image comply with the specified constraints.
+
+```xml
+<number_of_objects number_of_objects="187" mean_object_size="111.87" eps="0.01" exclude_labels="0" />
````
The other assertion use to have a syntax like:

```
<has_n_lines n=".."/>
```

or 

```
<has_size size="..."/>
```

Maybe we can get closer to this?

kostrykin replied:
I think assertions like `has_image_width` or `has_image_center_of_mass` are rather unintuitive[^1], but I agree that this scheme is more consistent with what is currently present[^2] in Galaxy. So I guess I will change it.

[^1]: In spoken language, you want to assert that *"the image has a width"* and *"the image has a center of mass"*, not that the output *"has an image center"* of something.
[^2]: https://docs.galaxyproject.org/en/latest/dev/schema.html#assertions-for-archives-zip-tar

kostrykin replied:
Done this in 00485ccc5294815c06e1d4d51d92d487f4133c65.

bernt-matthias replied:
Sorry for the delay and nitpicking: What I meant here is that it would be even more consistent with the present assertions to have 

```
<has_image_width witdth="" ...
```

instead of 

```
<has_image_witdth value=""
```

We [recently](https://github.com/galaxyproject/galaxy/pull/17490) also implemented this for `has_size`. 

Apart from this I'm super happy with the PR.

kostrykin replied:
Done this in 5964f72dfcb3092ecb3eac1037daa6dae099b2c0 and 8a81593c5372d1e2b6031e6ae62a820f4ef6857c.

Thanks for your suggestions!

bernt-matthias commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -2739,6 +2751,123 @@ $attribute_list::5
+        <xs:documentation xml:lang="en">The required number of unique labels in the image. It is assumed that each individual object corresponds to a unique label.</xs:documentation>
+      </xs:annotation>
+    </xs:attribute>
+    <xs:attribute name="mean_object_size" type="xs:float" use="optional">
```
Wondering if it would be better (easier to understand and test) to have separate assertions for the number and mean? Even if it's a bit more to write for the tool developers (they could use macros).

We could then formulate min, max, and delta for each of the two properties.

kostrykin replied:
Done this in 00485ccc5294815c06e1d4d51d92d487f4133c65.

bernt-matthias commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -2739,6 +2751,123 @@ $attribute_list::5
+        <xs:documentation xml:lang="en">The absolute tolerance to be used for the ``mean_object_size`` assertion (defaults to 0.01).</xs:documentation>
+      </xs:annotation>
+    </xs:attribute>
+    <xs:attribute name="exclude_labels" type="xs:string" use="optional">
```
Could it be useful to have `include_labels` as well?

kostrykin replied:
Not sure what the use case for this would be?

bernt-matthias replied:
Neither do I :) Might just be more convenient if the number of labels is large.

kostrykin replied:
Ah, you mean `include_labels` as in *only* include those labels. Gotcha.

kostrykin replied:
Done this in 00485ccc5294815c06e1d4d51d92d487f4133c65. The attribute is named simply `labels`, I think `include_labels` could be confusing (at least it confused me, like *"what's the point of including a label that is not present in the image?"*).

bernt-matthias commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -2739,6 +2751,123 @@ $attribute_list::5
+    </xs:attribute>
+    <xs:attribute name="eps" type="xs:float" use="optional" default="0.01">
+      <xs:annotation>
+        <xs:documentation xml:lang="en">The absolute tolerance to be used for the ``mean_intensity`` and ``center_of_mass`` assertions (defaults to ``0.01``).</xs:documentation>
```
Wondering also here if we should split this into two assertion, see my comment on `number_of_objects`.

kostrykin replied:
Done this in 00485ccc5294815c06e1d4d51d92d487f4133c65.

bernt-matthias commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
````diff
@@ -2739,6 +2751,123 @@ $attribute_list::5
+Asserts the image has the specified metadata.
+
+```xml
+<image_has_metadata width="512" height="512" channels="3" />
````
Maybe split into `<has_image_width>`, `<has_image_height>`, `<has_image_channels>` and add `min`, `max` and `delta` attributes?

kostrykin replied:
Done this in 00485ccc5294815c06e1d4d51d92d487f4133c65.

kostrykin committed:
Squashed commit of the following:

commit 33b660ee19fa58e84643bba0b875c874a7d3da12
Author: Leonid Kostrykin <void@evoid.de>
Date:   Tue Mar 5 20:58:24 2024 +0000

    Fix XSD

commit 4ee6112b84d3406d6ba98dd190fe07fda2d90307
Author: Leonid Kostrykin <void@evoid.de>
Date:   Tue Mar 5 20:53:14 2024 +0000

    Add `labels` attribute for `has_image_labels` and `has_image_mean_object_size`

commit 4ff401c4f0731cc3aceb9c511f7c3c61abf2b19c
Author: Leonid Kostrykin <void@evoid.de>
Date:   Tue Mar 5 20:32:57 2024 +0000

    Replace `image_has_labels` by `has_image_labels` and `has_image_mean_object_size`

commit 09149d1394a1b81c8c3036316ab2be355db24cde
Author: Leonid Kostrykin <void@evoid.de>
Date:   Tue Mar 5 19:51:56 2024 +0000

    Fix bug

commit 28fedb98f19a35ab6a4747aa5b435c329277df50
Author: Leonid Kostrykin <void@evoid.de>
Date:   Tue Mar 5 19:44:12 2024 +0000

    Replace `image_has_intensities` by `has_image_mean_intensity` and `has_image_center_of_mass`

commit fe96916ca28752f5c5512658db53c4162d8d7d66
Author: Leonid Kostrykin <void@evoid.de>
Date:   Tue Mar 5 19:10:38 2024 +0000

    Replace `image_has_metadata` by `has_image_width`, etc.

kostrykin committed:
Rename `delta` to `eps` for float-based assertions (#1)

* Rename `delta` to `eps` for float-based assertions

* Fix tests

kostrykin committed:
Rename assertion attribute `value`

kostrykin committed:
Rename `has_image_labels` to `has_image_n_labels`

bernt-matthias commented on lib/galaxy/tool_util/verify/asserts/image.py:
```diff
@@ -0,0 +1,289 @@
+    output_bytes: bytes,
+    channel: Optional[Union[int, str]] = None,
+    exclude_labels: Optional[Union[str, List[int]]] = None,
+    value: Optional[Union[int, str]] = None,
```
```suggestion
    n: Optional[Union[int, str]] = None,
```

bernt-matthias commented on lib/galaxy/tool_util/verify/asserts/image.py:
```diff
@@ -0,0 +1,289 @@
+    present_labels = _get_image_labels(output_bytes, channel, exclude_labels)[1]
+    _assert_number(
+        len(present_labels),
+        value,
```
```suggestion
        n,
```

kostrykin committed:
Update lib/galaxy/tool_util/verify/asserts/image.py

Co-authored-by: M Bernt <m.bernt@ufz.de>

kostrykin committed:
Update lib/galaxy/tool_util/verify/asserts/image.py

Co-authored-by: M Bernt <m.bernt@ufz.de>

kostrykin committed:
Fix linting issues
